### DISCUSSION TITLE
Question about auto-restart and multi-language implementation

### DISCUSSION THREAD

**Discussion by alx3dev** [fyne-io/fyne#3538] [Q&A] [answered]
Hi guys, I am pretty much new to Fyne, and not an expert in Golang (not a pure beginner, thought). If this is not a good place to ask a question like this, my apologize.

I wrote a simple implementation for my app to support multiple languages, but application require restart to change everything. This is not a problem for users, because they will mostly set it once, but would be great to restart app automatically. Is there any way to achieve that? Whatever I try, I can close it, but never start again.

Here is what I did:

I defined a new package called `translation` with `Translate` struct that contain words I want to translate.
```go
package translation

import (
	"fyne.io/fyne/v2"
)

type Translation struct {
	Home           string
	Settings       string
	Timer          string
	Pwm            string
}
```

Now I have methods `Initialize` that is called at start, and method `Set` to change language.
```go
func (t *Translation) Initialize() {
	language :=
		fyne.CurrentApp().Preferences().StringWithFallback("LANGUAGE", "English")

	t.Set(language)
}

func (t *Translation) Set(l string) {
	fyne.CurrentApp().Preferences().SetString("LANGUAGE", l)

	switch l {
	case "English":
		t.English()
	case "Serbian":
		t.Serbian()
	}
}
```

At the end, there are methods to add strings based on selected language:
```go
func (t *Translation) English() {
	t.Home = "Home"
	t.Settings = "Settings"
	t.Timer = "Time  "
	t.PWM = "PWM"
}

func (t *Translation) Serbian() {
	t.Home = "Насловна"
	t.Settings = "Подешавања"
	t.Timer = "Време"
	t.PWM = "Пулс  "
}
```
To use it, I put everything inside a struct:
```go
type MyAPP struct {
  APP fyne.APP
  WIN fyne.Window
  TR   translation.Translation
}

func StartApp() *MyApp {
  a := app.NewWithID('some-id')
  w := a.NewWindow('title')

  myapp := &MyApp{APP: a WIN: w}
  myapp.TR := translations.Translation{}
  myapp.TR.Initialize()

  return myapp
}
```  

Now in other methods I pass `myapp` pointer
```go
func doSomething(m *MyApp) {
  T := m.TR
  msgTab := container.NewTabItem(T.Home, container.NewPadded(some_function(f)))
  ...
  ...
}
```
Is there anything wrong with setup like this, and is there a way to restart app after selecting language? Language is selected from dropdown menu.

Edit: I will upload full source code tomorrow and update question with link, just in case someone want to take a look.

Edit 2: Link to the repository: https://github.com/alx3dev/Wireless-UV-Station

**Comment by Jacalz** [comment, marked as answer]
Hi there. Welcome. There is nothing wrong with asking here. You are more than welcome to do so :)

I think asking the application to restart itself is going to be quite hard to get working across platforms; or will at least require platform specific solutions for most of them. I suppose https://stackoverflow.com/questions/41376092/golang-how-to-auto-restart-process-when-binary-updated could be a useful place to start.


### Translations
As for if it is a good solution for translation, I don’t really have any huge input to give. It generally seems quite clean and well designed. I guess it probably could be useful as an external package for other Fyne application developers to use until we provide a built-in solution that integrates better into the toolkit.

The only thing that I could think of is that it might be easier for translators to implement translations by loading from files and not having to set fields individually. Both have their drawbacks though. Yours seems like it could be quite fast and avoid all the trouble of bundling files into the binary.

**Reply by Jacalz** [reply]
One extra note: It is faster to use your application variable directly (if you have access to it and it makes sense to pass it in) than to call `fyne.CurrentApp()`.

If you don’t want to pass it in, it is of course entirely fair to do as you have done.

**Reply by alx3dev** [reply]
I will take a look at link you provided. I agree it's a better to do that from file, but my app is a pretty simple one, so I wanted to avoid bundling, and yes it's really fast. It's also a pain in the ass for bigger applications :)

So for restart, I will need to build another package that will be triggered on app exit, so it can start an app again. This should be easy for desktops, but not so for mobile. I guess I will just provide users notification to restart app :)
Thank you on fast reply.

Edit: It's not a problem to pass a variable, i will do that.

**Comment by andydotxyz** [comment]
I would add to the answer above that you don't have to restart an application - you can add a listener to the preferences and each widget can update when the language changes. This is much faster for the end-user as restarting software is very slow in comparison to just updating the widgets on screen.

**Reply by alx3dev** [reply]
I will try to implement that, I've tried but failed and didn't wanted to waste my time. Generally problem was to update app-tabs names. Here is the link to my app, you can take a look if you want. Notification to user to restart app is there just to test system notifications.

https://github.com/alx3dev/Wireless-UV-Station

**Reply by andydotxyz** [reply]
You can set the `Text` field on each of the `TabItem`s in the container.

**Reply by alx3dev** [reply]
Thanks on advice, I saved pointers to tabs in a struct, so I can call `Text` on language change. Also I added bindings for control words `Timer`, `Power`, and `Speed`, now only dropdown menu itself is left, but I think I will leave it for now, it's not so important. In future I will probably just add icons with flags to solve it.

Only thing left is keyboard, so I can dynamically send commands over wifi or serial. Works on PC, but android is buggy - can't see what I am typing,  Is there a way to see entry field (single or multiline entry) while typing, or should I implement custom keyboard in app?

**Reply by andydotxyz** [reply]
There is an open issue for auto-resizing the window so Android /iOS does not hide content when it appears (#566).
Until then the workaround is to add padding to the bottom of the view and scroll down, sorry.
